h2/src/main/org/h2/util/DateTimeFunctions.java

```diff
@@ -374,6 +379,109 @@ public static Value extract(String part, Value value) {
         return result;
     }
 
+    /**
+     * Truncate the given date to 'day'
+     *
+     * @param timeUnit the time unit (e.g. 'DAY', 'HOUR', etc.)
+     * @param value the date
+     * @return date truncated to 'day'
+     */
+    public static Value truncateDate(String timeUnitStr, Value value) {
+        Value result;
+        int timeUnit = getDatePart(timeUnitStr);
+
+        // Retrieve the dateValue and the time in nanoseconds if the date.
+        long[] fieldDateAndTime = DateTimeUtils.dateAndTimeFromValue(value);
+        long dateValue = fieldDateAndTime[0];
+        long timeNanosRetrieved = fieldDateAndTime[1];
+
+        // Variable used to the time in nanoseconds of the date truncated.
+        Long timeNanos = null;
+
+        // Compute the number of time unit in the date, for example, the
+        // number of time unit 'HOUR' in '15:14:13' is '15'. Then convert the
+        // result to nanoseconds.
+        switch (timeUnit) {
+
+        case MICROSECOND:
+
+            long nanoInMicroSecond = 1_000l;
+            long microseconds = timeNanosRetrieved / nanoInMicroSecond;
+            timeNanos = microseconds * nanoInMicroSecond;
+            break;
+
+        case MILLISECOND:
+
+            long nanoInMilliSecond = 1_000_000l;
+            long milliseconds = timeNanosRetrieved / nanoInMilliSecond;
+            timeNanos = milliseconds * nanoInMilliSecond;
+            break;
+
+        case SECOND:
+
+            long nanoInSecond = 1_000_000_000l;
+            long seconds = timeNanosRetrieved / nanoInSecond;
+            timeNanos = seconds * nanoInSecond;
+            break;
+
+        case MINUTE:
+
+            long nanoInMinute = 60_000_000_000l;
+            long minutes = timeNanosRetrieved / nanoInMinute;
+            timeNanos = minutes * nanoInMinute;
+            break;
+
+        case HOUR:
+
+            long nanoInHour = 3_600_000_000_000l;
+            long hours = timeNanosRetrieved / nanoInHour;
+            timeNanos = hours * nanoInHour;
+            break;
+
+        case DAY_OF_MONTH:
+
+            timeNanos = 0l;
+            break;
+
+        case WEEK:
+
+            // Use Calendar to retrieve the first day of the week and convert
+            // the date to a dateValue
+            Date currentDate = DateTimeUtils.convertDateValueToDate(dateValue);
+            Calendar calendar = Calendar.getInstance();
+            calendar.setTime(currentDate);
+            calendar.set(Calendar.DAY_OF_WEEK, Calendar.MONDAY);
+            ValueTimestamp valueTimestamp = ValueTimestamp.fromMillis(calendar.getTimeInMillis());
```

Comment:
Use `DateTimeUtils.getIsoDayOfWeek()`. It returns 1 for Monday and so on. After this if day is not Monday (if it is Monday there is nothing to do any more) use something like
```Java
DateTimeUtils.dateValueFromAbsoluteDay(
    DateTimeUtils.absoluteDayFromDateValue(dateValue) - dayOfWeek + 1)
```
to get a new `dateValue`.

But check what day of week is used by PostgreSQL in different locales. May be this should be locale-specific.

Reply:
Or you can use a more optimized version like:
```Java
long absoluteDay = DateTimeUtils.absoluteDayFromDateValue(dateValue);
// The next method is private now
int dayOfWeek = DateTimeUtils.getDayOfWeekFromAbsolute(absoluteDay, /* Monday */ 1);
if (dayOfWeek != 1) {
    dateValue = dateValueFromAbsoluteDay(absoluteDay - dayOfWeek + 1);
}
```
It should be faster if `dayOfWeek` is not Monday, because in this case only one slow invocation of `DateTimeUtils.absoluteDayFromDateValue()` will be performed.

Reply:
Should I have to change the visibility of `DateTimeUtils.getDayOfWeekFromAbsolute` or use the `DateTimeUtils.getDayOfWeek` function:
``` java
long absoluteDay = DateTimeUtils.absoluteDayFromDateValue(dateValue);
int dayOfWeek = DateTimeUtils.getDayOfWeek(dateValue, 1);
// The next method is private now: we should not used it ?
// int dayOfWeek = DateTimeUtils.getDayOfWeekFromAbsolute(absoluteDay, 1);
if (dayOfWeek != 1) {
    dateValue = DateTimeUtils.dateValueFromAbsoluteDay(absoluteDay - dayOfWeek + 1);
}
``` 

Maybe it's better to change the visibility of the function `DateTimeUtils.getDayOfWeekFromAbsolute` in order to avoid to have two invocations of this function ?

Reply:
Yes, it's better to change visibility of this method.

There is no sense to invoke `absoluteDayFromDateValue()` two times for the same `dateValue`.